File: src/conn.py
#!/usr/bin/python
"""IRC Server Socket Connection Class"""

import socket
import time

import consts

def _check_conn(func):
    """Decorator to check that a connection is established before doing anything

    Args:
        func -- Function object to wrap
    Returns:
        Wrapped function
    """
    def wrap(*args, **kwargs):
        if not args[0].is_connected:
            raise Exception
        return func(*args, **kwargs)
    return wrap

class SocketConnection():
    """Manages the socket connection"""
    def __init__(self):
        """Constructor."""
        self._RX_BUF_SZ = 2048
        self._sock = None
        self.is_connected = False

    def connect(self):
        """Initializes connection to the Twitch servers."""
        if not self.is_connected:
            self._sock = socket.socket()
            self._sock.settimeout(0.5)
            self._sock.connect((consts.HOST, consts.PORT))
            self.is_connected = True

            self.send("PASS " + consts.PASS)
            self.send("NICK " + consts.USER)
            for chan in consts.TARGET_CHANNELS:
                print("Joining: " + chan)
                self.send("JOIN #" + chan)

    @_check_conn
    def send(self, msg):
        """Send a message on the socket. Will attempt reconnection if socket conn is lost.

        Args:
            msg -- string to send
        """
        try:
            self._sock.sendall((msg + "\r\n").encode("utf-8"))
        except socket.error as ex:
            print(ex)
            self.is_connected = False
            time.sleep(1)
            self.connect()

    @_check_conn
    def recv(self):
        """Read data from the socket. Timeout every 0.5 seconds.

        Returns:
            stripped decoded string or None if no data was recved
        """
        data = None
        try:
            data = self._sock.recv(self._RX_BUF_SZ).decode("utf-8").strip()
        except socket.timeout:
            pass
        return data

    def chat(self, channel, message):
        """Send a chat message on a channel

        Args:
            channel -- channel to send on
            message -- string to send
        """
        msg = "PRIVMSG #" + channel + " :" + message
        print("TX: " + msg)
        self.send(msg)

File: src/conn.go
package main

import (
	"fmt"
	"net"
	"strings"
	"time"
)

const (
	ircHostURL  = "irc.twitch.tv"
	ircHostPort = "6667"
)

type IrcConnection struct {
	conn        net.Conn
	isConnected bool
	connList    []string
	txQueue chan string
	control chan bool
}

// Returns a new IRC Client
func NewIRCConnection(conns []string) *IrcConnection {
	return &IrcConnection{
		conn:        nil,
		isConnected: false,
		connList:    conns,
		txQueue: make(chan string, 100),
		control: make(chan bool),
	}
}

// Connect to the IRC server, authenticate and join target channels
func (c *IrcConnection) Connect(nick string, pass string) error {
	if !c.isConnected {
		conn, err := net.Dial("tcp", ircHostURL+":"+ircHostPort)
		if err != nil {
			fmt.Println(err)
			return err
		}
		c.conn = conn
		go c.rateLimiter()
		c.send("PASS " + pass)
		c.send("NICK " + nick)
		for _, channel := range c.connList {
			c.Join(channel)
		}
		c.isConnected = true
	}
	return nil
}

func (c *IrcConnection) Join(channel string) {
	c.send("JOIN #" + channel)
}

func (c *IrcConnection) Part(channel string) {
	c.send("PART #" + channel)
}

// Disconnect from the IRC server
func (c *IrcConnection) Disconnect() error {
	if c.isConnected {
		c.control <- true
		err := c.conn.Close()
		if err != nil {
			fmt.Println(err)
			return err
		}
		c.isConnected = false
	}
	return nil
}

// Send a chat message to an IRC channel
func (c *IrcConnection) Chat(channel string, message string) {
	c.txQueue <- "PRIVMSG #" + channel + " : " + message

}

// Rate limit the transmission of messages to the IRC server
func (c *IrcConnection) rateLimiter() {
	for {
		select {
        case <-c.control:
            return
        case msg := <-c.txQueue:
            c.send(msg)
        }
        time.Sleep(2 * time.Second)
	}
}

// Receieve data from the IRC connection. Handles ping pong automatically.
func (c *IrcConnection) Recv() (string, error) {
	buf := make([]byte, 4096)
	len, err := c.conn.Read(buf)
	if err != nil {
		c.Disconnect()
		return "", err
	}
	// Cast to string, trim newlines
	message := strings.TrimSpace(string(buf[:len]))

	// Handle ping pong, return the next recv
	if message == "PING :tmi.twitch.tv" {
		c.send("PONG :tmi.twitch.tv")
		message, err = c.Recv()
		if err != nil {
			return "", err
		}
	}
	fmt.Println("RX: " + message)
	return message, nil
}

// Raw send to the socket
func (c *IrcConnection) send(message string) {
	fmt.Fprintf(c.conn, message+"\r\n")
	fmt.Println("TX: " + message)
}
